await-signals/await_signals_workflow.go

```diff
@@ -0,0 +1,164 @@
+ *  and rules around order of their arrivals and timeouts increases.
+ *
+ *  The following example demonstrates an alternative approach. It receives signals in a separate goroutine.
+ *  Each signal handler just updates a correspondent shared variable with the signal data.
+ *  The main workflow function awaits the next step using `workflow.AwaitWithTimeout` using condition composed of
+ *  the shared variables. This makes the main workflow method free from signal callbacks and makes the business logic
+ *  clear.
+ */
+
+// SignalToSignalTimeout is them maximum time between signals
+var SignalToSignalTimeout = 30 * time.Second
+
+// FromFirstSignalTimeout is the maximum time to receive all signals
+var FromFirstSignalTimeout = 60 * time.Second
+
+type AwaitSignals struct {
+	FirstSignalTime time.Time
+	Signal1Received bool
+	Signal2Received bool
+	Signal3Received bool
+}
+
+// Listen to signals Signal1, Signal2, and Signal3
+func (a *AwaitSignals) Listen(ctx workflow.Context) {
+	log := workflow.GetLogger(ctx)
+	for {
+		selector := workflow.NewSelector(ctx)
+		selector.AddReceive(workflow.GetSignalChannel(ctx, "Signal1"), func(c workflow.ReceiveChannel, more bool) {
+			c.Receive(ctx, nil)
+			a.Signal1Received = true
+			log.Info("Signal1 Received")
+		})
+		selector.AddReceive(workflow.GetSignalChannel(ctx, "Signal2"), func(c workflow.ReceiveChannel, more bool) {
+			c.Receive(ctx, nil)
+			a.Signal2Received = true
+			log.Info("Signal2 Received")
+		})
+		selector.AddReceive(workflow.GetSignalChannel(ctx, "Signal3"), func(c workflow.ReceiveChannel, more bool) {
+			c.Receive(ctx, nil)
+			a.Signal3Received = true
+			log.Info("Signal3 Received")
+		})
+		selector.Select(ctx)
+		if a.FirstSignalTime.IsZero() {
+			a.FirstSignalTime = workflow.Now(ctx)
+		}
+	}
+}
+
+// GetNextTimeout returns the maximum time allowed to wait for the next signal.
+func (a *AwaitSignals) GetNextTimeout(ctx workflow.Context) (time.Duration, error) {
+	if a.FirstSignalTime.IsZero() {
+		panic("FirstSignalTime is not yet set")
+	}
+	total := workflow.Now(ctx).Sub(a.FirstSignalTime)
+	totalLeft := FromFirstSignalTimeout - total
+	if totalLeft <= 0 {
+		return 0, temporal.NewApplicationError("FromFirstSignalTimeout", "timeout")
+	}
+	return common.MinDuration(SignalToSignalTimeout, totalLeft), nil
+}
+
+// AwaitSignalsWorkflow workflow definition
+func AwaitSignalsWorkflow(ctx workflow.Context) (err error) {
+	log := workflow.GetLogger(ctx)
+	var a AwaitSignals
+	workflow.Go(ctx, func(ctx workflow.Context) {
+		a.Listen(ctx)
+	})
+
+	// Wait for Signal1
+	err = workflow.Await(ctx, func() bool {
+		return a.Signal1Received
+	})
+	// Cancellation
+	if err != nil {
+		return
+	}
+	log.Info("Signal1 Processed")
+
+	// Wait for Signal2
+	timeout, err := a.GetNextTimeout(ctx)
+	// No time left. At this point this cannot really happen.
+	if err != nil {
+		return
+	}
+	ok, err := workflow.AwaitWithTimeout(ctx, timeout, func() bool {
```

Comment:
Also a good use case for your new `ReceiveWithTimeout`, but I understand that doesn't demonsrtate what you want here